src/ui/components/post/post_listing/thumbnail.rs

```diff
@@ -0,0 +1,112 @@
+use crate::{
+  ui::components::common::icon::{Icon, IconSize, IconType},
+  utils::{is_image, is_video},
+};
+use lemmy_client::lemmy_api_common::lemmy_db_schema::newtypes::PostId;
+use leptos::*;
+use leptos_router::A;
+use std::rc::Rc;
+use tailwind_fuse::{AsTailwindClass, TwVariant};
+
+#[derive(TwVariant)]
+enum ThumbnailIconType {
+  #[tw(default, class = "m-auto")]
+  NoImage,
+  #[tw(class = "rounded-bl bg-slate-600/75 block text-white/75 absolute right-0 top-0")]
+  Image,
+}
+
+#[derive(TwVariant)]
+#[tw(class = "border-2 border-neutral size-20 rounded-lg p-0 grid-in-thumbnail")]
+enum ThumbnailWrapperType {
+  #[tw(default, class = "flex")]
+  NoImage,
+  #[tw(class = "relative overflow-hidden")]
+  Image,
+}
+
+#[derive(Clone)]
+struct ThumbnailData {
+  pub icon: IconType,
+  pub icon_size: IconSize,
+  pub wrapper_class: Rc<str>,
+  pub icon_class: Rc<str>,
+}
+
+#[component]
+pub fn Thumbnail(
+  url: Memo<Option<Rc<str>>>,
+  image_url: Memo<Option<Rc<str>>>,
+  #[prop(into)] has_embed_url: Signal<bool>,
+  id: PostId,
+) -> impl IntoView {
+  let thumbnail_data = Signal::derive(move || {
+    with!(|url, image_url, has_embed_url| {
+      let (icon_size, icon_class, wrapper_class) = if image_url.is_some() {
```

Comment:
For all the if cases in this function, it'd be good to add a comment.

Reply:
Added coments.